dashboard/app/assets/javascripts/header.js

```diff
@@ -51,36 +55,8 @@ dashboard.buildHeader = function (stageData, progressData, currentLevelId, userI
   if (progressData.linesOfCodeText) {
     $('.header_popup .header_text').text(progressData.linesOfCodeText);
   }
-  var progressContainer = $('.progress_container');
-  var serverProgress = progressData.levels || {};
-  stageData.levels.forEach(function(level, index, levels) {
-    var status;
-    if (dashboard.clientState.queryParams('user_id')) {
-      // Show server progress only (the student's progress)
-      status = activityCssClass((serverProgress[level.id] || {}).result);
-    } else {
-      // Merge server progress with local progress
-      status = mergedActivityCssClass((serverProgress[level.id] || {}).result, clientProgress[level.id]);
-    }
-    var defaultClass = level.kind == 'assessment' ? 'puzzle_outer_assessment' : 'puzzle_outer_level';
-    var href = level.url;
-    if (userId) {
-      href += '?user_id=' + userId;
-    }
-    if (sectionId) {
-      href += '&section_id=' + sectionId;
-    }
-    var link = $('<a>').attr('id', 'header-level-' + level.id).attr('href', href).addClass('level_link').addClass(status).text(level.title);
 
-    if (level.kind == 'unplugged') {
-      link.addClass('unplugged_level');
-    }
-    var div = $('<div>').addClass(level.id === currentLevelId ? 'puzzle_outer_current' : defaultClass).append(link);
-    if (index === levels.length - 1) {
-      div.addClass('last');
-    }
-    progressContainer.append(div).append('\n');
-  });
+  renderStageProgress(stageData, progressData, clientProgress, currentLevelId);
```

Comment:
renderStageProgress might also end up making more sense living on the new namespace dashboard.progress?